apps/champions/lib/champions/battle/simulator.ex

```diff
@@ -63,187 +64,459 @@ defmodule Champions.Battle.Simulator do
-    new_value = calculate_value(effect, caster, target)
-    Map.put(target, String.to_atom(effect.stat_affected), new_value)
+  # Apply an effect to its target *if it hits*.
+  # Hit chance is affected by the ChanceToApply component and could be expanded later on.
+  # Returns the new state of the target.
+  defp maybe_apply_effect(effect, target, caster, current_step_number) do
+    if effect_hits?(effect) do
+      apply_effect(effect, target, caster, current_step_number)
+    else
+      Logger.info("#{format_unit_name(effect.caster)}'s effect missed.")
+      target
+    end
   end
 
-  defp calculate_value(
-         %{amount_format: "additive", stat_based_on: nil} = effect,
-         _caster,
-         target
-       ),
-       do: target[String.to_atom(effect.stat_affected)] + effect.amount
+  # Return whether an effect with a ChanceToApply component hits.
+  # Later on, this might also handle similar mechanics like the target's dodge chance.
+  defp effect_hits?(effect) do
+    chance_to_apply_component =
+      Enum.find(effect.components, fn comp -> comp["type"] == "ChanceToApply" end)
 
-  defp calculate_value(
-         %{amount_format: "multiplicative", stat_based_on: nil} = effect,
-         _caster,
-         target
-       ),
-       do: target[String.to_atom(effect.stat_affected)] * effect.amount
+    case chance_to_apply_component do
+      nil ->
+        true
 
-  defp calculate_value(
-         %{amount_format: "additive", stat_based_on: stat_based_on} = effect,
-         caster,
-         target
-       ),
-       do:
-         target[String.to_atom(effect.stat_affected)] +
-           floor(effect.amount * caster[String.to_atom(stat_based_on)] / 100)
+      chance_to_apply_component ->
+        chance_to_apply_component["chance"] >= :rand.uniform()
+    end
+  end
 
+  # Apply an effect to its target. Returns the new state of the target.
+  # For now this applies the executions on the spot.
+  # Later on, it will "cast" them as we do with skills and effects to account for execution delays.
+  # Returns the new state of the target.
+  defp apply_effect(effect, target, caster, current_step_number) do
+    target_after_modifiers =
+      Enum.reduce(effect.modifiers, target, fn modifier, target ->
+        # If it's permanent, we set its duration to -1
+        new_modifier =
+          modifier
+          |> Map.put(:remaining_steps, Map.get(effect.type, "duration", -1))
+          |> Map.put(:step_applied_at, current_step_number)
+
+        case modifier.modifier_operation do
+          "Add" ->
+            put_in(target, [:modifiers, :additives], [new_modifier | target.modifiers.additives])
+
+          "Multiply" ->
+            put_in(target, [:modifiers, :multiplicatives], [new_modifier | target.modifiers.multiplicatives])
+
+          "Override" ->
+            put_in(target, [:modifiers, :overrides], [new_modifier | target.modifiers.overrides])
+        end
```

Comment:
You're putting all the modifiers into the target but later on the `calculate_unit_stat` uses the modifiers from the caster instead of the target

Reply:
I think there might be a misunderstanding here. calculate_unit_stat is not used to set new values, it's used to get how the stat of a unit looks like right now, when you factor in the modifiers that affect it. It's used, for example, when calculating how much the attack_ratio of a DealDamage execution should be multiplied by. The comment above calculate_unit_stat "Calculate the current amount of the given attribute that the unit has, based on its modifiers." aims to explain this. Let me kwow if you think I should add any more comments to make this clearer for future reviewers/mantainers.

Reply:
hmmm no, i think i was in the wrong here not understanding. After reading it again i get it now!